packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,425 @@
+     * to the Optimistic Oracle on proposed distribution.
+     * @param optimisticOracleLivenessTime Liveness period in seconds during which proposed distribution can be
+     * disputed through Optimistic Oracle.
+     */
+    function createReward(
+        uint256 maximumRewardAmount,
+        uint256 earliestProposalTimestamp,
+        uint256 optimisticOracleProposerBond,
+        uint256 optimisticOracleLivenessTime,
+        bytes32 priceIdentifier,
+        IERC20 rewardToken,
+        bytes calldata customAncillaryData
+    ) external nonReentrant() {
+        require(address(merkleDistributor) != address(0), "Missing MerkleDistributor");
+        require(_getIdentifierWhitelist().isIdentifierSupported(priceIdentifier), "Identifier not registered");
+        require(_ancillaryDataWithinLimits(customAncillaryData), "Ancillary data too long");
+        require(optimisticOracleLivenessTime >= MINIMUM_LIVENESS, "OO liveness too small");
+        require(optimisticOracleLivenessTime < MAXIMUM_LIVENESS, "OO liveness too large");
+
+        // Pull maximum rewards from the sponsor.
+        rewardToken.safeTransferFrom(msg.sender, address(this), maximumRewardAmount);
+
+        // Store funded reward and log created reward.
+        Reward memory reward =
+            Reward({
+                sponsor: msg.sender,
+                rewardToken: rewardToken,
+                maximumRewardAmount: maximumRewardAmount,
+                earliestProposalTimestamp: earliestProposalTimestamp,
+                optimisticOracleProposerBond: optimisticOracleProposerBond,
+                optimisticOracleLivenessTime: optimisticOracleLivenessTime,
+                priceIdentifier: priceIdentifier,
+                customAncillaryData: customAncillaryData
+            });
+        uint256 rewardIndex = nextCreatedReward;
+        rewards[rewardIndex] = reward;
+        emit RewardCreated(
+            reward.sponsor,
+            reward.rewardToken,
+            rewardIndex,
+            reward.maximumRewardAmount,
+            reward.earliestProposalTimestamp,
+            reward.optimisticOracleProposerBond,
+            reward.optimisticOracleLivenessTime,
+            reward.priceIdentifier,
+            reward.customAncillaryData
+        );
+
+        // Bump nextCreatedReward index.
+        nextCreatedReward++;
+    }
+
+    /**
+     * @notice Allows anyone to deposit additional rewards for distribution before `earliestProposalTimestamp`.
+     * @dev The caller must approve this contract to transfer `additionalRewardAmount` amount of `rewardToken`.
+     * @param rewardIndex Index for identifying existing Reward struct that should receive additional funding.
+     * @param additionalRewardAmount Additional reward amount that the sponsor is posting for distribution.
+     */
+    function increaseReward(uint256 rewardIndex, uint256 additionalRewardAmount) external nonReentrant() {
```

Comment:
Potential reward recipients should be able to start their work (e.g. voting in bribing rewards) already when initial rewards are funded. In case of irreversible on-chain votes this could be a problem if promised reward is decreased after the fact.